### ISSUE TITLE
 Add support for service upgrades without configuration files in rancher-compose

### ISSUE THREAD

**Issue by nelsonsilva** [rancher/rancher#3571]
Even if it's doable through the REST API (https://github.com/rancher/rancher/issues/1874#issuecomment-180082502) `rancher-compose` should support doing in-service upgrades without the configuration files.

**Comment by silviupanaite** [comment]
+ 1

**Comment by deniseschannon** [comment]
How are we expecting Rancher-compose to upgrade? Are we just expecting all the services in the stack to be upgraded?

The only thing that I could see changing is that it would re-pull the image if you have selected the appropriate label (`io.rancher.container.pull_image:always`).

**Comment by nelsonsilva** [comment]
Yes, we would just need to rebuild the container, re-pulling the image if requested. Sometimes just rebuilding the container with the same image is enough given that the entrypoint would be run again and could itself do whatever is necessary to update.

**Comment by adamkdean** [comment]
I would see the workflow as, for a particular service, not stack:

```
docker build -t quay.io/username/myservice .
docker push quay.io/username/myservice
rancher-compose --upgrade myservice
run-some-tests myservice
rancher-compose --confirm-upgrade myservice (or rancher-compose --rollback myservice)
```

Obviously I've generalised but, this is the workflow I'd like to see at least.

**Comment by adamkdean** [comment]
You do not want to see the deployment script I just wrote to handle this.

http://i.imgur.com/nQ50U.gif

**Comment by deniseschannon** [comment]
https://forums.rancher.com/t/using-rancher-cli-to-do-upgrades-without-dealing-with-docker-compose-yml/4187

**Comment by obenziane2** [comment]
I have the same issue here, we use tags on the our images , example: `docker build -t example.com/username/myservice:v2  .`    be able to update the image or tag will be great.

**Comment by qtzar** [comment]
I would also be very interesting in seeing this capability for use in my CI/CD flow.
Something like
rancher upgrade stack/service -forcepull

**Comment by ddelnano** [comment]
We have been developing our own cli for upgrades and cloning entire environments at my work place.  I just made the code and docker image available.  You can find it [here](https://github.com/nowait-tools/rancher-cli)

**Comment by broth-eu** [comment]
Is there any progress on this feature?
My workaround is currently to directly working with the underlying docker containers which is quite complex.

**Comment by OliDow** [comment]
Has there been any updates on this feature being added to rancher CLI?
I have been using rancher-compose but after adding secrets to my compose file rancher-compose is very unhappy indeed.

**Comment by vetalpaprotsky** [comment]
I would like to have this kind of functionality too. It will make my CI/CD flow much easier.

**Comment by jwstric2** [comment]
Agreed with most of the above comments; there are so many projects spinning up that are attempting to mediate this problem with accessing the APIs (https://github.com/etlweather/gaucho and https://github.com/nowait-tools/rancher-cli as example ... they both work well by the way do not dismissing these projects at all)

From a CD perspective; we ended up adding the best of both worlds to leverage rancher-compose as it's rancher supported and not have a docker-compose / rancher-compose to sync in our repository:

1.  Access the APIs to download the docker-compose and rancher compose data (dockerCompose and rancherCompose in the json response).  Example:  projects/1a5/stacks/1st42.  
2.  Then use rancher compose as normal with the data; substituing the new image:tag for the upgrade
3.  Run some tests
4.  Confirm upgrade with compose

I've yet to had a chance to look at the rancher-compose code to see if a workflow would be easy to add to it for runtime config download with substitution and be acceptable to the community of users.  Time is against me right now :)

**Comment by jonaskello** [comment]
I use [cdrx/rancher-gitlab-deploy](https://github.com/cdrx/rancher-gitlab-deploy) which works great in gitlab. However native support in the rancher cli would of course be nice.

**Comment by cr0hn** [comment]
I just released a smalls script that allow to upgrade Rancher services. I hope it could the useful.

Here the URL of the project:

https://github.com/cr0hn/rancher-upgrader

**Comment by hernandev** [comment]
haha that's funny, I had a different case for this.

@cr0hn I also made one, like 1 year ago: https://github.com/codecasts/rancher-deployer

Not even sure if it still works since I've been using gitlab-deployer, but they kinda do the same.

I'm revisiting this since I'm moving out Gitlab

**Comment by zot24** [comment]
Is there any way to achieve this for Rancher 1.6.X? 

Is it a feature that we might see in the future for Rancher CLI? or will Rancher CLI be somehow deprecated and we'll see something different more in line with kubectl for Rancher 2.0?

Thanks!

**Comment by kimaero** [comment]
Another easy workaround is to do something like this using Rancher CLI:
```
mkdir /tmp/rancher_stack_name && cd /tmp/rancher_stack_name
rancher export stack_name && cd stack_name
rancher up -d --pull --upgrade --confirm-upgrade service_name another_service_name
rm -rf /tmp/rancher_stack_name
```
